Sources/Plasma/FeatureLib/pfConsole/pfConsole.cpp

```diff
@@ -525,145 +527,76 @@ void    pfConsole::IHandleKey( plKeyEventMsg *msg )
-                    biglines[0] = 0;
-                    for ( i=fPythonMultiLines; i>0 ; i--)
-                    {
-                        // reach back in the history and find this line and paste it in here
-                        int recall = fHistory[ fPythonMode ].fCursor - i;
-                        if ( recall < 0 )
-                            recall += kNumHistoryItems;
-                        strcat(biglines,fHistory[ fPythonMode ].fData[ recall ]);
-                        strcat(biglines,"\n");
-                    }
-                    // now evaluate this mess they made
-                    PyObject* mymod = PythonInterface::FindModule("__main__");
-                    PythonInterface::RunStringInteractive(biglines,mymod);
-                    std::string output;
-                    // get the messages
-                    PythonInterface::getOutputAndReset(&output);
-                    AddLine( output.c_str() );
-                    // all done doing multi lines...
-                    fPythonMultiLines = 0;
-                }
-            }
-            // else we are just doing single lines
-            else
-            {
-                // was there actually anything in the input buffer?
-                if ( fWorkingLine[0] != 0 )
-                {
-                    AddLineF(">>> %s", fWorkingLine);
-                    // check to see if this is going to be a multi line mode ( a ':' at the end)
-                    if ( fWorkingLine[eol-1] == ':' )
-                    {
-                        fPythonMultiLines = 1;
-                    }
-                    else
-                    // else if not the start of a multi-line then execute it
-                    {
-                        PyObject* mymod = PythonInterface::FindModule("__main__");
-                        PythonInterface::RunStringInteractive(fWorkingLine,mymod);
-                        std::string output;
-                        // get the messages
-                        PythonInterface::getOutputAndReset(&output);
-                        AddLine( output.c_str() );
-                    }
-                }
-                else
-                    AddLine( ">>> " );
-            }
-            // find the end of the line
-            for( c = fWorkingLine, eol = 0; *c != 0; eol++, c++ );
-        }
-        else
-        {
-            if( !fEngine->RunCommand( fWorkingLine, IAddLineCallback ) )
-            {
-                c = (char *)fEngine->GetErrorMsg();
-                if( c[ 0 ] != 0 )
-                    AddLine( c );
-            }
-        }
-
-        // Clear
-        fWorkingLine[ fWorkingCursor = 0 ] = 0;
+        IExecuteWorkingLine();
         findAgain = false;
         findCounter = 0;
         IUpdateTooltip();
     }
     else if( msg->GetKeyCode() == KEY_END )
     {
-        fWorkingCursor = strlen( fWorkingLine );
+        fWorkingCursor = strnlen(fWorkingLine, std::size(fWorkingLine));
```

Comment:
I did some reading up on these functions (they are C functions, not C++, right?) and wondered if `strnlen_s` might be better, as it also returns 0 if `fWorkingLine` is null and `strnlen` will not do so? I was reading from: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strnlen-strnlen-s?view=msvc-170

Reply:
`strlen_s` is a Microsoft-only function in the MSVCRT and cannot be used on non-Windows platforms (unless you build with something esoteric, like winelib).